ethers-etherscan/src/account.rs

```diff
@@ -0,0 +1,532 @@
+}
+
+/// The pre-defined block parameter for balance API endpoints
+pub enum Tag {
+    Earliest,
+    Pending,
+    Latest,
+}
+
+impl ToString for Tag {
+    fn to_string(&self) -> String {
+        match self {
+            Tag::Earliest => "earliest",
+            Tag::Pending => "pending",
+            Tag::Latest => "latest",
+        }
+        .to_string()
+    }
+}
+
+impl Default for Tag {
+    fn default() -> Self {
+        Tag::Latest
+    }
+}
+
+/// The list sorting preference
+pub enum Sort {
+    Asc,
+    Desc,
+}
+
+impl ToString for Sort {
+    fn to_string(&self) -> String {
+        match self {
+            Sort::Asc => "asc",
+            Sort::Desc => "desc",
+        }
+        .to_string()
+    }
+}
+
+/// Common optional arguments for the transaction or event list API endpoints
+pub struct TxListParams {
+    start_block: u64,
+    end_block: u64,
+    page: u64,
+    offset: u64,
+    sort: Sort,
+}
+
+impl TxListParams {
+    pub fn new(start_block: u64, end_block: u64, page: u64, offset: u64, sort: Sort) -> Self {
+        Self { start_block, end_block, page, offset, sort }
+    }
+}
+
+impl Default for TxListParams {
+    fn default() -> Self {
+        Self { start_block: 0, end_block: 99999999, page: 0, offset: 10000, sort: Sort::Asc }
+    }
+}
+
+impl From<TxListParams> for HashMap<&'static str, String> {
+    fn from(tx_params: TxListParams) -> Self {
+        let mut params = HashMap::new();
+        params.insert("startBlock", tx_params.start_block.to_string());
+        params.insert("endBlock", tx_params.end_block.to_string());
+        params.insert("page", tx_params.page.to_string());
+        params.insert("offset", tx_params.offset.to_string());
+        params.insert("sort", tx_params.sort.to_string());
+        params
+    }
+}
+
+/// Options for querying internal transactions
+pub enum InternalTxQueryOption<A: AsRef<str>> {
+    ByAddress(A),
+    ByTransactionHash(A),
+    ByBlockRange,
+}
+
+/// Options for querying ERC20 or ERC721 token transfers
+pub enum TokenQueryOption<A: AsRef<str>> {
+    ByAddress(A),
+    ByContract(A),
+    ByAddressAndContract(A, A),
+}
+
+impl<A: AsRef<str>> TokenQueryOption<A> {
+    pub fn into_params(self, list_params: TxListParams) -> HashMap<&'static str, String> {
+        let mut params: HashMap<&'static str, String> = list_params.into();
+        match self {
+            TokenQueryOption::ByAddress(address) => {
+                params.insert("address", address.as_ref().to_string());
+                params
+            }
+            TokenQueryOption::ByContract(contract) => {
+                params.insert("contractaddress", contract.as_ref().to_string());
+                params
+            }
+            TokenQueryOption::ByAddressAndContract(address, contract) => {
+                params.insert("address", address.as_ref().to_string());
+                params.insert("contractaddress", contract.as_ref().to_string());
+                params
```

Comment:
looks like we're always allocating via `as_ref().to_string()`

we could alternatively make `A: Into<Cow>` so that we don't need to alloc a new string if an address is given as &str